File: euler19.py
# Сколько воскресений выпадает на первое число месяца в двадцатом веке
# (с 1 января 1901 года до 31 декабря 2000 года)?

from datetime import datetime
from calendar import monthrange


def isleap_year(year: int) -> bool:
    """
    Проверяет високосный ли год.

    Args:
        year: Год для проверки.

    Returns:
        True если год високосный, иначе False.
    """
    return year % 400 == 0 or (year % 4 == 0 and year % 100 != 0)


def counter_prime_weekday(year: int, weeknum: int) -> int:
    """
    Возвращает количество совпадений заданного дня недели с 1 числом месяца.

    Args:
        year: Конечный год периода (с 1900).
        weeknum: Номер дня недели (1-7).

    Returns:
        Количество совпадений.
    """
    assert year >= 1900 and 1 <= weeknum <= 7
    month30 = [4, 6, 9, 11]
    febrary = {'standard': 28, 'leap': 29}
    days_month, sum_days = [], []
    for y in range(1900, year + 1):
        year_type = 'leap' if isleap_year(y) else 'standard'
        for month in range(1, 13):
            if month in month30:
                days_month.append(30)
            elif month == 2:
                days_month.append(febrary[year_type])
            else:
                days_month.append(31)
        sum_days += days_month
        days_month = []
    alldays = counter_prime = 0
    for x in sum_days:
        alldays += x
        if (alldays + 1) % weeknum == 0:
            counter_prime += 1
    return counter_prime


def euler19() -> None:
    """
    Решение задачи Эйлера №19.

    Сколько воскресений выпадает на первое число месяца в двадцатом веке?
    """
    n = 2000

    # 1 вариант
    start_time = datetime.now()
    print(counter_prime_weekday(n, 7) - counter_prime_weekday(1900, 7))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    counter_prime = 0
    for year in range(1901, n + 1):
        for month in range(1, 13):
            if monthrange(year, month)[0] == 6:
                counter_prime += 1
    print(counter_prime)
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler19()

File: go/euler19.go
// euler19 считает количество воскресений на первое число месяца
// в двадцатом веке.
//
// Решение задачи Эйлера №19.
package main

import (
	"fmt"
	"time"
)

func isLeapYear(year int) bool {
	return year%400 == 0 || (year%4 == 0 && year%100 != 0)
}

func counterPrimeWeekday(year int, weeknum int) int {
	month30 := []int{4, 6, 9, 11}
	sumDays := 0
	counterPrime := 0
	for y := 1900; y <= year; y++ {
		daysMonth := make([]int, 0)
		for month := 1; month <= 12; month++ {
			if contains(month30, month) {
				daysMonth = append(daysMonth, 30)
			} else if month == 2 {
				if isLeapYear(y) {
					daysMonth = append(daysMonth, 29)
				} else {
					daysMonth = append(daysMonth, 28)
				}
			} else {
				daysMonth = append(daysMonth, 31)
			}
		}
		sumDays += sumSlice(daysMonth)
		daysMonth = nil
	}
	allDays := 0
	for _, x := range sumDays {
		allDays += x
		if (allDays+1)%weeknum == 0 {
			counterPrime++
		}
	}
	return counterPrime
}

func contains(slice []int, val int) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func sumSlice(slice []int) int {
	total := 0
	for _, v := range slice {
		total += v
	}
	return total
}

func euler19() {
	n := 2000

	startTime := time.Now()
	fmt.Println(counterPrimeWeekday(n, 7) - counterPrimeWeekday(1900, 7))
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	counterPrime := 0
	for year := 1901; year <= n; year++ {
		for month := 1; month <= 12; month++ {
			day := firstDayOfMonth(year, month)
			if day == 6 { // Sunday (0=Monday in Python's weekday())
				counterPrime++
			}
		}
	}
	fmt.Println(counterPrime)
	fmt.Println(time.Since(startTime))
}

func firstDayOfMonth(year, month int) int {
	// Using Zeller's congruence
	if month < 3 {
		month += 12
		year--
	}
	k := year % 100
	j := year / 100
	h := (1 + (13*(month+1))/5 + k + k/4 + j/4 + 5*j) % 7
	return (h + 6) % 7
}

func main() {
	euler19()
}
